File: 55-jump-game/jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] == 0) {
                int j = i - 1;
                while (j >= 0 && nums[j] <= i - j)
                    --j;
                if (j == -1)
                    return false;
                i = j;
            }
        }
        return true;
    }
}

File: 55-jump-game/jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        arr = [0 for i in range(len(nums) + 1)]
        for i in range(len(nums)):
            arr[min(i + 1, len(nums))] += 1
            arr[min(i + nums[i] + 1, len(nums))] -= 1
        for i in range(1, len(arr) - 1):
            arr[i] += arr[i - 1]
            if arr[i] == 0:
                return False
        return True